rfcs/rework-index-updates.md

```diff
@@ -0,0 +1,60 @@
+# Summary
+
+This RFC proposes moving away from the current model of fetching new releases
+to build, moving from running [crates-index-diff] in a thread to using webhooks
+and the [crates-index] crate.
+
+# Motivation
+
+While the current approach has worked well for us so far, it has some problems:
+
+* Running the update in a cronjob every 2 minutes is wasteful, as there is
+  often a greater delay between two publishes.
+* Running the update in a crobjob every 2 minutes adds delay to getting the
+  documentation built if the queue is empty, as the release might potentially
+  have to wait those extra two minutes.
+* The approach doesn't scale, if we want to move to a setup where there is more
+  than a single frontend server we'd have to elect which server runs the fetch.
+* The way crates-index-diff stores its state (a branch in the local repo) is
+  fragile, as it might become out of sync causing the loss of a publish.
+* The way crates-index-diff stores its state makes it hard to move the server
+  installation, as the index repository needs to be moved as well.
```

Comment:
Given #766 being implemented, I think it's fine to discard the state and start the new server from current `HEAD` when initialized. If the new server starts up while the old server is still running so that the old server queues stuff up until the `HEAD` of the new server then there will be no missed crates, otherwise a consistency check can be run afterwards to queue missed crates.

Reply:
If we end up running docs.rs on ECS (which is a thing I'd *love* to see) starting a new server from scratch might happen every day or even more often (every time we deploy and every time ECS feels like doing that), so we could actually miss a lot of crates while we wait for synchronization.

Reply:
Ok. Another alternative would be for us to store an extra copy of the last seen state in the database ourselves, and use that to initialize a new server.

Reply:
I like the idea of state being stored in the database instead of on the host. I think it would make it easier to have multiple build servers eventually: instead of each server having to have it's own fully synchronized index, each build server is stateless and builds crates when it gets a (sufficiently authenticated) webhook. Then, we only need one version of the index that's synchronized on the database, and if we find during synchronization that we missed a crate, we can just send out a webhook ourselves.